JSON.mqh

```diff
@@ -90,24 +90,298 @@ class JSON {
+        key = NULL;
+      }
+      else
+      if (ch == ']')
+      {
+        if (expectingKey || expectingValue || current.GetType() != JsonNodeArray) {
+          return GracefulReturn("Unexpected end of array", i, root, key);
+        }
+        
+        current = current.GetParent();
+        expectingValue = false;
+      }
+      else
+      if (ch >= '0' && ch <= '9')
+      {
+        if (!expectingValue) {
+          return GracefulReturn("Unexpected numeric value", i, root, key);
+        }
+        
+        if (!ExtractNumber(data, i, extracted)) {
+          return GracefulReturn("Cannot parse numberic value", i, root, key);
+        }
+        
+        value = StringFind(extracted, ".") != -1 ? JsonParam::FromValue(StringToDouble(extracted)) : JsonParam::FromValue(StringToInteger(extracted));
+        current.AddChild(new JsonNode(current.GetType() == JsonNodeObject ? JsonNodeObjectProperty : JsonNodeArrayItem, current, key, value));
+        expectingValue = false;
+        
+        // Skipping value.
+        i += StringLen(extracted) - 1;
+        
+        // We don't want to delete it twice.
+        key = NULL;
+      }
+      else
+      if (ch == '"')
+      {
+        // A string value.
+        if (!expectingValue) {
+          return GracefulReturn("Unexpected quotes", i, root, key);
+        }
+        
+        extracted = ExtractString(data, i+1);
+        
+        if (extracted == NULL) {
+          return GracefulReturn("Unexpected end of file when parsing string", i, root, key);
+        }
+        
+        // Skipping double quotes.
+        i += StringLen(extracted) + 1;
+                  
+        current.AddChild(new JsonNode(current.GetType() == JsonNodeObject ? JsonNodeObjectProperty : JsonNodeArrayItem, current, key, JsonParam::FromString(extracted)));
+        
+        expectingValue = false;
+       
+        // We don't want to delete key twice.
+        key = NULL;
+      }
+      else
+      if (ch == ',') {
+        if (expectingKey || expectingValue || expectingSemicolon) {
+          return GracefulReturn("Unexpected comma", i, root, key);
+        }
+        
+        if (current.GetType() == JsonNodeObject)
+          expectingKey = true;
+        else
+          expectingValue = true;
+      }
+    }
+    
+    if (key)
+      delete key;
+            
+    return root;
+  }
+  
+  static JsonNode* GracefulReturn(string error, unsigned int index, JsonNode* root, JsonParam* key) {
+    Print(error + " at index ", index);
+  
+    if (root != NULL)
+      delete root;
+      
+    if (key != NULL)
+      delete key;
+
+    return NULL;
+  }
+  
+  static bool ExtractNumber(string& data, unsigned int index, string& number) {
+    string str;
+     
+    for (unsigned int i = index; i < (unsigned int)StringLen(data); ++i)
+    {
+    #ifdef __MQL5__
+      unsigned short ch = StringGetCharacter(data, i);
+    #else
+      unsigned short ch = StringGetChar(data, i);
```

Comment:
It seems MT4 also supports [`StringGetCharacter()`](https://docs.mql4.com/strings/stringgetcharacter).